File: 02_prs_weights.R
#!/usr/bin/env Rscript
# ============================================================================
# Step 2: PRS Weight Calculation using GWAS Summary Statistics
# ============================================================================
# This script fetches GWAS Catalog summary statistics for target diseases,
# calculates polygenic risk scores for each 1000 Genomes individual, and
# exports scores for downstream ML training.
# ============================================================================

suppressPackageStartupMessages({
  library(data.table)
  library(yaml)
  library(curl)
})

ROOT <- dirname(sys.frame(1)$ofile)
cfg <- yaml::read_yaml(file.path(ROOT, "config.yaml"))

PROJECT <- cfg$gcp$project_id
DATASET <- cfg$bigquery$dataset
BUCKET <- cfg$gcp$bucket

# ---------------------------------------------------------------------------
# 1. Load GWAS Catalog summary statistics from public FTP
# ---------------------------------------------------------------------------
fetch_gwas_sumstats <- function(efo_id, disease_name) {
  cat(sprintf("Fetching GWAS Catalog stats for: %s (%s)\n", disease_name, efo_id))

  base_url <- "https://www.ebi.ac.uk/gwas/rest/api/studies"
  url <- sprintf("%s/search/findByEfo?efo=%s", base_url, efo_id)

  resp <- tryCatch(
    curl::curl_fetch_memory(url),
    error = function(e) {
      cat(sprintf("  WARN: GWAS API fetch failed: %s\n", e$message))
      return(NULL)
    }
  )

  if (is.null(resp)) return(NULL)

  data <- jsonlite::fromJSON(rawToChar(resp$content))
  if (length(data$_embedded$studies) == 0) {
    cat("  No studies found\n")
    return(NULL)
  }

  # Get the first study with summary stats
  study <- data$_embedded$studies[1, ]
  cat(sprintf("  Study: %s\n", study$authorShort))

  # Fetch SNP associations
  assoc_url <- sprintf("%s/studies/%s/associations",
                       base_url, study$accession)
  assoc_resp <- curl::curl_fetch_memory(assoc_url)
  assoc_data <- jsonlite::fromJSON(rawToChar(assoc_resp$content))

  if (length(assoc_data$_embedded$associations) == 0) return(NULL)

  assocs <- assoc_data$_embedded$associations
  results <- data.table()

  for (i in seq_len(nrow(assocs))) {
    row <- assocs[i, ]

    if (is.null(row$riskAlleles) || length(row$riskAlleles) == 0) next

    for (j in seq_along(row$riskAlleles)) {
      allele <- row$riskAlleles[[j]]
      if (is.null(allele$riskFrequency) || is.null(allele$effects)) next

      for (k in seq_along(allele$effects)) {
        eff <- allele$effects[[k]]
        beta_val <- tryCatch(as.numeric(eff$beta), error = function(e) NA_real_)
        pval <- tryCatch(as.numeric(eff$pvalue), error = function(e) NA_real_)
        or_val <- tryCatch(as.numeric(eff$oddsRatio), error = function(e) NA_real_)

        # Convert OR to log-odds if beta missing
        if (is.na(beta_val) && !is.na(or_val)) beta_val <- log(or_val)

        results <- rbind(results, data.table(
          rsid = allele$riskAllele$variation$rsID %||% NA_character_,
          risk_allele = allele$riskAllele$riskAllele,
          beta = beta_val,
          pvalue = pval,
          or = or_val,
          freq = as.numeric(allele$riskFrequency),
          disease = disease_name
        ))
      }
    }
  }

  results <- results[!is.na(beta) & !is.na(pvalue) & pvalue > 0]
  results <- unique(results, by = "rsid")

  cat(sprintf("  Found %d SNP-effect pairs\n", nrow(results)))
  return(results)
}

`%||%` <- function(a, b) if (!is.null(a)) a else b


# ---------------------------------------------------------------------------
# 2. Read genotype matrix from local parquet (exported by Step 1)
# ---------------------------------------------------------------------------
read_genotype_matrix <- function() {
  cat("Reading genotype matrix...\n")
  gmat_path <- file.path(ROOT, "genotype_matrix.parquet")

  if (!file.exists(gmat_path)) {
    stop("genotype_matrix.parquet not found. Run 01_variant_extraction.py first.")
  }

  gmat <- arrow::read_parquet(gmat_path)
  cat(sprintf("  Loaded %d samples x %d variants\n", nrow(gmat), ncol(gmat) - 1))
  return(gmat)
}


# ---------------------------------------------------------------------------
# 3. Calculate PRS per individual per disease
# ---------------------------------------------------------------------------
calculate_prs <- function(gmat, gwas_stats, disease_key) {
  cat(sprintf("Calculating PRS for: %s\n", disease_key))

  variant_col <- names(gmat)[1]
  sample_ids <- gmat[[variant_col]]
  snp_cols <- names(gmat)[-1]

  # Match variants between genotype matrix and GWAS stats
  matched <- merge(
    data.table(rsid = snp_cols),
    gwas_stats[, .(rsid, beta, pvalue)],
    by = "rsid",
    all.x = FALSE
  )

  cat(sprintf("  Matched %d / %d variants with GWAS stats\n",
              nrow(matched), length(snp_cols)))

  if (nrow(matched) == 0) {
    warning(sprintf("No variants matched for %s", disease_key))
    return(data.table(sample_id = sample_ids, prs = NA_real_,
                      disease = disease_key))
  }

  # PRS = sum(dosage_i * beta_i)
  dosages <- as.matrix(gmat[, matched$rsid, with = FALSE])
  betas <- matched$beta

  prs <- dosages %*% betas
  prs <- as.numeric(prs)

  result <- data.table(
    sample_id = sample_ids,
    prs = prs,
    disease = disease_key,
    n_snps = nrow(matched)
  )

  cat(sprintf("  PRS range: [%.2f, %.2f], mean: %.2f\n",
              min(prs), max(prs), mean(prs)))

  return(result)
}


# ---------------------------------------------------------------------------
# 4. Population stratification: compute PCs using flashpca2
# ---------------------------------------------------------------------------
compute_pcs <- function(gmat, n_pcs = 10) {
  cat("Computing population PCs via flashpca2...\n")

  variant_col <- names(gmat)[1]
  sample_ids <- gmat[[variant_col]]
  dosages <- as.matrix(gmat[, -1, with = FALSE])

  # Write temporary PLINK files for flashpca2
  tmp_prefix <- file.path(tempdir(), "tmp_pca")

  # .bed file
  bed <- matrix(as.raw(dosages), nrow = nrow(dosages))
  writeBin(as.vector(bed), paste0(tmp_prefix, ".bed"))

  # .bim file
  bim <- data.table(
    chr = 1, snp = colnames(dosages),
    pos = seq_along(colnames(dosages)),
    pos2 = seq_along(colnames(dosages)),
    A1 = "A", A2 = "G"
  )
  fwrite(bim, paste0(tmp_prefix, ".bim"), sep = "\t", col.names = FALSE)

  # .fam file
  fam <- data.table(
    fam1 = 1, iid = sample_ids,
    fid = 0, pid = 0, sex = 0, pheno = -9
  )
  fwrite(fam, paste0(tmp_prefix, ".fam"), sep = "\t", col.names = FALSE)

  # Run flashpca2
  pc_file <- paste0(tmp_prefix, ".pcs.txt")
  cmd <- sprintf("flashpca2 --bfile %s --ndim %d --out %s 2>&1",
                 tmp_prefix, n_pcs, pc_file)

  result <- tryCatch(
    system(cmd, intern = TRUE),
    error = function(e) {
      cat("  flashpca2 not available, falling back to SVD\n")
      return(NULL)
    }
  )

  if (!is.null(result) && file.exists(pc_file)) {
    pcs <- fread(pc_file)
    cat(sprintf("  Computed %d PCs\n", ncol(pcs) - 1))
    return(pcs)
  }

  # Fallback: PCA via SVD
  cat("  Using SVD-based PCA\n")
  scaled <- scale(dosages, center = TRUE, scale = FALSE)
  svd_result <- svd(scaled, nu = n_pcs, nv = 0)
  pc_matrix <- svd_result$u[, 1:n_pcs] * svd_result$d[1:n_pcs]

  pcs <- data.table(sample_id = sample_ids, as.data.table(pc_matrix))
  names(pcs) <- c("sample_id", paste0("PC", 1:n_pcs))

  return(pcs)
}


# ---------------------------------------------------------------------------
# 5. Main execution
# ---------------------------------------------------------------------------
main() {
  cat("=" |> rep(60) |> paste(collapse = ""), "\n")
  cat("Step 2: PRS Weight Calculation\n")
  cat("=" |> rep(60) |> paste(collapse = ""), "\n\n")

  # Load genotype data
  gmat <- read_genotype_matrix()

  # Fetch GWAS stats for each disease
  all_prs <- list()

  for (disease_key in names(cfg$diseases)) {
    disease <- cfg$diseases[[disease_key]]
    efo_id <- disease$gwas_efo

    gwas_stats <- fetch_gwas_sumstats(efo_id, disease$name)

    if (is.null(gwas_stats) || nrow(gwas_stats) == 0) {
      # Fallback: use curated candidate variants from config
      cat(sprintf("  Using curated candidate variants for %s\n", disease$name))
      gwas_stats <- rbindlist(lapply(disease$variants, function(v) {
        data.table(
          rsid = v$rsid,
          risk_allele = "A",
          beta = 0.3,   # placeholder effect sizes
          pvalue = 1e-5,
          or = 1.35,
          freq = 0.3,
          disease = disease$name
        )
      }))
    }

    prs <- calculate_prs(gmat, gwas_stats, disease_key)
    all_prs[[disease_key]] <- prs
  }

  # Combine PRS across diseases
  combined_prs <- rbindlist(all_prs)

  # Compute ancestry PCs
  pcs <- compute_pcs(gmat, n_pcs = cfg$prs$n_pcs)

  # Save PRS scores
  out_dir <- file.path(ROOT, "output")
  dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)

  fwrite(combined_prs, file.path(out_dir, "prs_scores.csv"))
  fwrite(pcs, file.path(out_dir, "population_pcs.csv"))

  cat(sprintf("\nSaved PRS scores: %d rows\n", nrow(combined_prs)))
  cat(sprintf("Saved PCs: %d samples x %d components\n", nrow(pcs), cfg$prs$n_pcs))

  # Upload to GCS
  system(sprintf(
    "gsutil cp %s/gs://%s/prs_scores.csv",
    file.path(out_dir, "prs_scores.csv"),
    BUCKET
  ))
  system(sprintf(
    "gsutil cp %s/gs://%s/population_pcs.csv",
    file.path(out_dir, "population_pcs.csv"),
    BUCKET
  ))

  cat("\nDone. Next: Run 03_vertex_prs_train.py\n")
}

main()

File: 02_prs_weights.py
"""Step 2: PRS Weight Calculation using GWAS Catalog Summary Statistics.

Fetches GWAS associations from the GWAS REST API for each disease,
calculates polygenic risk scores for 1000 Genomes individuals,
and computes population PCs for stratification adjustment.
"""

import json
import logging
from pathlib import Path

import numpy as np
import pandas as pd
import requests
import yaml
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

ROOT = Path(__file__).parent
with open(ROOT / "config.yaml") as f:
    cfg = yaml.safe_load(f)

OUTPUT_DIR = ROOT / "output"


# ---------------------------------------------------------------------------
# 1. Fetch GWAS summary statistics from GWAS Catalog REST API
# ---------------------------------------------------------------------------
def fetch_gwas_sumstats(efo_id: str, disease_name: str) -> pd.DataFrame:
    """Fetch GWAS Catalog association data for a given EFO trait."""
    logger.info(f"Fetching GWAS stats for: {disease_name} ({efo_id})")

    base_url = "https://www.ebi.ac.uk/gwas/rest/api"

    # Find studies associated with this EFO term
    study_url = f"{base_url}/studies/search/findByEfo?efo={efo_id}"
    try:
        resp = requests.get(study_url, timeout=30)
        resp.raise_for_status()
        studies = resp.json()["_embedded"]["studies"]
    except Exception as e:
        logger.warning(f"  GWAS API fetch failed: {e}")
        return pd.DataFrame()

    if not studies:
        logger.warning(f"  No studies found for {efo_id}")
        return pd.DataFrame()

    logger.info(f"  Found {len(studies)} studies")

    # Get associations from the first study
    study = studies[0]
    assoc_url = study["_links"]["associations"]["href"]

    try:
        resp = requests.get(assoc_url, timeout=30)
        resp.raise_for_status()
        associations = resp.json()["_embedded"]["associations"]
    except Exception as e:
        logger.warning(f"  Failed to fetch associations: {e}")
        return pd.DataFrame()

    records = []
    for assoc in associations:
        for risk_allele in assoc.get("riskAlleles", []):
            variation = risk_allele.get("variation", {})
            rsid = variation.get("rsID")
            if not rsid:
                continue

            for effect in risk_allele.get("effects", []):
                beta = effect.get("beta")
                pvalue = effect.get("pvalue")
                or_val = effect.get("oddsRatio")

                beta_f = float(beta) if beta else None
                pval_f = float(pvalue) if pvalue else None

                # Convert OR to log-odds if beta is missing
                if beta_f is None and or_val:
                    beta_f = np.log(float(or_val))

                if beta_f is not None and pval_f is not None and pval_f > 0:
                    records.append({
                        "rsid": rsid,
                        "risk_allele": risk_allele.get("riskAllele", ""),
                        "beta": beta_f,
                        "pvalue": pval_f,
                        "freq": float(risk_allele.get("riskFrequency", 0)),
                        "disease": disease_name,
                    })

    df = pd.DataFrame(records)
    if not df.empty:
        df = df.drop_duplicates(subset=["rsid"])
        logger.info(f"  Found {len(df)} SNP-effect pairs")
    return df


# ---------------------------------------------------------------------------
# 2. Use curated candidate variants as fallback
# ---------------------------------------------------------------------------
def get_candidate_variants() -> pd.DataFrame:
    """Build GWAS-like summary stats from curated config variants."""
    records = []
    for disease_key, disease in cfg["diseases"].items():
        for v in disease["variants"]:
            # Use realistic effect sizes from literature
            effect_sizes = {
                "rs429358": 0.45,    # APOE e4
                "rs10455872": 0.35,  # LPA
                "rs10757278": 0.25,  # 9p21
                "rs688": 0.15,       # LDLR
                "rs7903146": 0.30,   # TCF7L2
                "rs1801282": 0.20,   # PPARG
                "rs5219": 0.18,      # KCNJ11
                "rs9939609": 0.22,   # FTO
                "rs17782313": 0.15,  # MC4R
                "rs6457617": 0.50,   # HLA-DRB1
                "rs11209026": -0.40, # IL23R (protective)
                "rs3087243": 0.20,   # CTLA4
            }

            records.append({
                "rsid": v["rsid"],
                "risk_allele": "A",
                "beta": effect_sizes.get(v["rsid"], 0.2),
                "pvalue": 1e-6,
                "freq": 0.3,
                "disease": disease["name"],
            })

    return pd.DataFrame(records)


# ---------------------------------------------------------------------------
# 3. Load genotype matrix from Step 1
# ---------------------------------------------------------------------------
def load_genotype_matrix() -> pd.DataFrame:
    """Load the dosage matrix from parquet."""
    path = OUTPUT_DIR / "genotype_matrix.parquet"
    if not path.exists():
        raise FileNotFoundError(f"Run 01_variant_extraction.py first: {path}")
    gmat = pd.read_parquet(path).reset_index()
    logger.info(f"Loaded genotype matrix: {gmat.shape}")
    return gmat


# ---------------------------------------------------------------------------
# 4. Calculate PRS per individual per disease
# ---------------------------------------------------------------------------
def calculate_prs(gmat: pd.DataFrame, gwas_stats: pd.DataFrame, disease_key: str) -> pd.DataFrame:
    """PRS = sum(dosage_i * beta_i) for matched variants."""
    logger.info(f"Calculating PRS for: {disease_key}")

    sample_ids = gmat["sample_id"]
    snp_cols = [c for c in gmat.columns if c != "sample_id"]

    # Match variants
    matched = gwas_stats[gwas_stats["rsid"].isin(snp_cols)]
    logger.info(f"  Matched {len(matched)} / {len(snp_cols)} variants with GWAS stats")

    if matched.empty:
        logger.warning(f"  No variants matched for {disease_key}")
        return pd.DataFrame({"sample_id": sample_ids, "prs": 0.0, "disease": disease_key})

    dosages = gmat[matched["rsid"].values].values
    betas = matched["beta"].values

    prs = dosages @ betas

    result = pd.DataFrame({
        "sample_id": sample_ids.values,
        "prs": prs,
        "disease": disease_key,
        "n_snps": len(matched),
    })

    logger.info(f"  PRS range: [{prs.min():.2f}, {prs.max():.2f}], mean: {prs.mean():.2f}")
    return result


# ---------------------------------------------------------------------------
# 5. Compute population PCs
# ---------------------------------------------------------------------------
def compute_pcs(gmat: pd.DataFrame, n_pcs: int = 10) -> pd.DataFrame:
    """Compute principal components from genotype matrix."""
    logger.info(f"Computing top {n_pcs} PCs...")

    snp_cols = [c for c in gmat.columns if c != "sample_id"]
    dosages = gmat[snp_cols].values.astype(float)

    # Standardize
    scaler = StandardScaler()
    dosages_scaled = scaler.fit_transform(dosages)

    pca = PCA(n_components=min(n_pcs, len(snp_cols)))
    pc_values = pca.fit_transform(dosages_scaled)

    pc_df = pd.DataFrame(pc_values, columns=[f"PC{i+1}" for i in range(pc_values.shape[1])])
    pc_df.insert(0, "sample_id", gmat["sample_id"].values)

    logger.info(f"  Explained variance: {pca.explained_variance_ratio_.sum():.2%}")
    return pc_df


# ---------------------------------------------------------------------------
# 6. Main
# ---------------------------------------------------------------------------
def main():
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    logger.info("=" * 60)
    logger.info("Step 2: PRS Weight Calculation (Python)")
    logger.info("=" * 60)

    # Load genotype matrix
    gmat = load_genotype_matrix()

    # Try GWAS Catalog API, fallback to curated variants
    all_prs = []

    for disease_key, disease in cfg["diseases"].items():
        gwas_stats = fetch_gwas_sumstats(disease["gwas_efo"], disease["name"])

        if gwas_stats.empty:
            logger.info(f"  Using curated candidate variants for {disease['name']}")
            gwas_stats = get_candidate_variants()
            gwas_stats = gwas_stats[gwas_stats["disease"] == disease["name"]]

        prs = calculate_prs(gmat, gwas_stats, disease_key)
        all_prs.append(prs)

    # Combine PRS across diseases
    combined_prs = pd.concat(all_prs, ignore_index=True)

    # Compute PCs
    pcs = compute_pcs(gmat, n_pcs=cfg["prs"]["n_pcs"])

    # Save
    combined_prs.to_csv(OUTPUT_DIR / "prs_scores.csv", index=False)
    pcs.to_csv(OUTPUT_DIR / "population_pcs.csv", index=False)

    logger.info(f"\nSaved PRS scores: {combined_prs.shape}")
    logger.info(f"Saved PCs: {pcs.shape}")

    # Summary
    logger.info("\nPRS Summary by Disease:")
    summary = combined_prs.groupby("disease")["prs"].agg(["count", "mean", "std", "min", "max"])
    logger.info(f"\n{summary.to_string()}")

    logger.info("\nDone. Next: Run 03_vertex_prs_train.py")


if __name__ == "__main__":
    main()
